src/api/substring.cc

```diff
@@ -0,0 +1,182 @@
+    Isolate* isolate = args.GetIsolate();
+    auto context = isolate->GetCurrentContext();
+    int argc = args.Length();
+
+    if (argc < SUBSTRING_MIN_ARGC) {
+        isolate->ThrowException(Exception::TypeError(
+                        String::NewFromUtf8(isolate,
+                        "Wrong number of arguments",
+                        NewStringType::kNormal).ToLocalChecked()));
+        return;
+    }
+
+    auto result = args[1];
+    if (argc == SUBSTRING_MIN_ARGC) {
+        args.GetReturnValue().Set(result);
+        return;
+    }
+
+
+    auto subject = args[2];
+    int resultLen = TO_V8STRING(result)->Length();
+    int subjectLen = TO_V8STRING(subject)->Length();
+    int start = TO_INTEGER_VALUE(args[3], context);
+    int end = subjectLen;
+    if (argc > 4) {
+        end = TO_INTEGER_VALUE(args[4], context);
+    }
+
+    if (resultLen == 0) {
+        args.GetReturnValue().Set(result);
+        return;
+    }
+    auto transaction = GetTransaction(GetLocalStringPointer(args[0]));
+    if (transaction == nullptr) {
+        args.GetReturnValue().Set(result);
+        return;
+    }
+
+    if (subjectLen <= 1) {
+        args.GetReturnValue().Set(result);
+        return;
+    }
+
+    auto taintedObj = transaction->FindTaintedObject(GetLocalStringPointer(subject));
+    if (!taintedObj) {
+        args.GetReturnValue().Set(result);
+        return;
+    }
+
+    try {
+        start = MAX(start, 0);
+        auto newRanges = getRangesInSlice(transaction, taintedObj, MIN(start, end), MAX(start, end));
+        if (newRanges && newRanges->Size() > 0) {
+            if (resultLen == 1) {
+                result = tainted::NewExternalString(isolate, args[1]);
+            }
+            transaction->AddTainted(GetLocalStringPointer(result), newRanges, result);
+        }
+    } catch (const std::bad_alloc& err) {
+    } catch (const container::QueuedPoolBadAlloc& err) {
+    } catch (const container::PoolBadAlloc& err) {
+    }
+
+    args.GetReturnValue().Set(result);
+}
+
+void substr(const FunctionCallbackInfo<Value>& args) {
+    Isolate* isolate = args.GetIsolate();
+    auto context = isolate->GetCurrentContext();
+    int argc = args.Length();
+
+    if (argc < SUBSTRING_MIN_ARGC) {
+        isolate->ThrowException(Exception::TypeError(
+                        String::NewFromUtf8(isolate,
+                        "Wrong number of arguments",
+                        NewStringType::kNormal).ToLocalChecked()));
+        return;
+    }
+
+    auto result = args[1];
+    int resultLen = TO_V8STRING(result)->Length();
+    if (argc == SUBSTRING_MIN_ARGC) {
+        args.GetReturnValue().Set(result);
+        return;
+    }
+
+    if (resultLen == 0) {
+        args.GetReturnValue().Set(result);
+        return;
+    }
+
+    auto subject = args[2];
+    int subjectLen = TO_V8STRING(subject)->Length();
+    int start = TO_INTEGER_VALUE(args[3], context);
+    int end = subjectLen;
```

Comment:
It would be better to rename `end` to `length` in order to match the original `substr` api